### PR TITLE
COMP: Address gcc4.8 warnings in itkContourExtractor2DImageFilter.hxx.

### PR DESCRIPTION
**Pull request by Leengit**
Attempting Issue #2395.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [X] Updated API documentation (or API not changed)
- [X] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [X] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [X] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKExamples) for all new major features (if any)

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx [hunk 1]
```diff
@@ -299,7 +299,7 @@ ContourExtractor2DImageFilter<TInputImage>::GenerateDataForLabels()
   {
     allLabels.assign(inputRange.cbegin(), inputRange.cend());
     std::sort(allLabels.begin(), allLabels.end());
-    const LabelsConstIterator last{ std::unique(allLabels.begin(), allLabels.end()) };
+    const LabelsIterator last{ std::unique(allLabels.begin(), allLabels.end()) };
```
Comment (N-Dekker) [inline Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:302]: Did the use of `LabelsConstIterator` really cause a GCC warning? If so, can you please mention which warning it is?

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx [hunk 2]
```diff
@@ -384,7 +384,7 @@ ContourExtractor2DImageFilter<TInputImage>::GenerateDataForLabels()
   mt->ParallelizeArray(
     0,
     allLabels.size(),
-    [this, &allLabels, &input, &labelsRegions, totalPixelCount, &labelsContoursOutput](SizeValueType i) {
+    [this, &allLabels, &input, &labelsRegions, totalPixelCount, &labelsContoursOutput](SizeValueType i) -> void {
```
Comment (N-Dekker) [inline Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:387]: Is it necessary for a clean compilation to add `-> void` here? If so, can you please mention the warning that it has solved?

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx (hunk 1 above)
Comment (Leengit) [reply Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:302]: https://open.cdash.org/viewBuildError.php?buildid=7093944 includes the following.  If I have misunderstood or if you see a better way, let's do that instead.
```plaintext
In file included from /home/kitware/Dashboards/TestsModules/Filtering/Path/include/itkContourExtractor2DImageFilter.h:318:0,
                 from /home/kitware/Dashboards/TestsModules/Filtering/Path/test/itkContourExtractor2DImageFilterTest.cxx:21:
/home/kitware/Dashboards/Tests/ITK/Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx: In instantiation of 'void itk::ContourExtractor2DImageFilter<TInputImage>::GenerateDataForLabels() [with TInputImage = itk::Image<unsigned char, 2u>]':
/home/kitware/Dashboards/TestsModules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:52:5:   required from 'void itk::ContourExtractor2DImageFilter<TInputImage>::GenerateData() [with TInputImage = itk::Image<unsigned char, 2u>]'
/home/kitware/Dashboards/TestsModules/Filtering/Path/test/itkContourExtractor2DImageFilterTest.cxx:758:1:   required from here
/home/kitware/Dashboards/TestsModules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:303:5: error: no matching function for call to 'std::vector<unsigned char, std::allocator<unsigned char> >::erase(const LabelsConstIterator&, std::vector<unsigned char, std::allocator<unsigned char> >::iterator)'
     allLabels.erase(last, allLabels.end());
     ^
/home/kitware/Dashboards/TestsModules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:303:5: note: candidates are:
In file included from /usr/include4.8/vector:69:0,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkConstShapedNeighborhoodIterator.h:21,
                 from /home/kitware/Dashboards/TestsModules/Filtering/Path/test/itkContourExtractor2DImageFilterTest.cxx:19:
/usr/include4.8/bits/vector.tcc:134:5: note: std::vector<_Tp, _Alloc>::iterator std::vector<_Tp, _Alloc>::erase(std::vector<_Tp, _Alloc>::iterator) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::iterator = __gnu_cxx::__normal_iterator<unsigned char*, std::vector<unsigned char, std::allocator<unsigned char> > >; typename std::_Vector_base<_Tp, _Alloc>::pointer = unsigned char*]
     vector<_Tp, _Alloc>::
     ^
/usr/include4.8/bits/vector.tcc:134:5: note:   candidate expects 1 argument, 2 provided
/usr/include4.8/bits/vector.tcc:146:5: note: std::vector<_Tp, _Alloc>::iterator std::vector<_Tp, _Alloc>::erase(std::vector<_Tp, _Alloc>::iterator, std::vector<_Tp, _Alloc>::iterator) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>; std::vector<_Tp, _Alloc>::iterator = __gnu_cxx::__normal_iterator<unsigned char*, std::vector<unsigned char, std::allocator<unsigned char> > >; typename std::_Vector_base<_Tp, _Alloc>::pointer = unsigned char*]
     vector<_Tp, _Alloc>::
     ^
/usr/include4.8/bits/vector.tcc:146:5: note:   no known conversion for argument 1 from 'const LabelsConstIterator {aka const __gnu_cxx::__normal_iterator<const unsigned char*, std::vector<unsigned char, std::allocator<unsigned char> > >}' to 'std::vector<unsigned char, std::allocator<unsigned char> >::iterator {aka __gnu_cxx::__normal_iterator<unsigned char*, std::vector<unsigned char, std::allocator<unsigned char> > >}'
In file included from /home/kitware/Dashboards/TestsModules/Filtering/Path/include/itkContourExtractor2DImageFilter.h:318:0,
                 from /home/kitware/Dashboards/TestsModules/Filtering/Path/test/itkContourExtractor2DImageFilterTest.cxx:21:
/home/kitware/Dashboards/TestsModules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:365:63: error: invalid initialization of reference of type 'const itk::ContourExtractor2DImageFilter<TInputImage>::GenerateDataForLabels() [with TInputImage = itk::Image<unsigned char, 2u>]::BoundingBoxType&' from expression of type '<brace-enclosed initializer list>'
       const BoundingBoxType & bbox{ labelBoundingBoxes[label] };
```

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx (hunk 2 above)
Comment (Leengit) [reply Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:387]: https://open.cdash.org/viewBuildError.php?buildid=7093944 includes the following.  If I have misunderstood or if you see a better way, let's do that instead.
```plaintext
In file included from /usr/include4.8/memory:79:0,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkExceptionObject.h:22,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkMacro.h:1341,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkLightObject.h:21,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkObject.h:31,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkRegion.h:31,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkImageRegion.h:31,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkImage.h:21,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkConstNeighborhoodIterator.h:24,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkNeighborhoodIterator.h:24,
                 from /home/kitware/Dashboards/TestsModules/Core/Common/include/itkConstShapedNeighborhoodIterator.h:23,
                 from /home/kitware/Dashboards/TestsModules/Filtering/Path/test/itkContourExtractor2DImageFilterTest.cxx:19:
/usr/include/c++/4.8/functional: In substitution of 'template<class _Res, class ... _ArgTypes> template<class _Functor> using _Invoke = decltype (std::__callable_functor(declval<_Functor&>())((declval<_ArgTypes>)()...)) [with _Functor = itk::ContourExtractor2DImageFilter<TInputImage>::GenerateDataForLabels() [with TInputImage = itk::Image<unsigned char, 2u>]::__lambda8; _Res = void; _ArgTypes = {long unsigned int}]':
/usr/include4.8/functional:2257:9:   required from 'void itk::ContourExtractor2DImageFilter<TInputImage>::GenerateDataForLabels() [with TInputImage = itk::Image<unsigned char, 2u>]'
/home/kitware/Dashboards/TestsModules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:52:5:   required from 'void itk::ContourExtractor2DImageFilter<TInputImage>::GenerateData() [with TInputImage = itk::Image<unsigned char, 2u>]'
/home/kitware/Dashboards/TestsModules/Filtering/Path/test/itkContourExtractor2DImageFilterTest.cxx:758:1:   required from here
/usr/include4.8/functional:2181:71: error: use of 'itk::ContourExtractor2DImageFilter<TInputImage>::GenerateDataForLabels() [with TInputImage = itk::Image<unsigned char, 2u>]::__lambda8' before deduction of 'auto'
  using _Invoke = decltype(__callable_functor(std::declval<_Functor&>())
                                                                       ^
/usr/include4.8/functional:2257: confused by earlier errors, bailing out
Preprocessed source stored into /tmp/ccJDVoSG.out file, please attach this to your bugreport.
```

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx (hunk 1 above)
Comment (N-Dekker) [reply Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:302]: Ah, thanks @Leengit so now I understand, the `std::vector` implementation of GCC 4.8 does not yet have the C++11 `erase(const_iterator)` overload.

Could you possibly adjust the commit message, to clarify that it also addresses errors, not just warnings? For example (remaining at <= maximum 72 characters):

> COMP: Fix gcc4.8 warnings + errors itkContourExtractor2DImageFilter.hxx

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx (hunk 2 above)
Comment (N-Dekker) [reply Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:387]: @Leengit Thanks for these details! I'm sorry I don't really understand what's going on here. I did try to reproduce the gcc4.8 error at https://godbolt.org/z/6o5enc but it compiles fine without an explicit "-> void". 

Looking forward to say goodbye to GCC 4 🤔

**Review by N-Dekker — APPROVED** [review]
It would be helpful if you could explicitly mention the GCC warning and error messages in the commit text. (Not necessarily the full output, of course. Just the most essential part of each GCC message would be fine to me.) But anyway, thanks, Lee! Hereby approved!

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx (hunk 1 above)
Comment (Leengit) [reply Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:302]: I changed the commit message.

**Pull request merged**